subprojects/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/failure/RootAssertionToFailureMapper.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing.failure;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.tasks.testing.TestFailure;
+
+/**
+ * Interface implemented by root-level classes which transform between {@link Throwable} and {@link TestFailure}.
+ * <p>
+ * As sometimes failure processing needs to be recursive (e.g. {@link org.gradle.api.internal.tasks.testing.failure.mappers.OpentestMultipleFailuresMapper}),
+ * they receive an instance implementing this interface to transform inner failures.
+ *
+ * @see org.gradle.api.internal.tasks.testing.junit.JUnitTestEventAdapter
+ * @see org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestExecutionListener
+ */
+@NonNullApi
+public interface RootAssertionToFailureMapper {
+
+    TestFailure createFailure(Throwable throwable);
```

Comment:
🤔 Would it make sense to return the list of mapper instances here?

Reply:
If your idea is to only supply the list of `FailureMapper` classes here: I think that exactly against the flexibility we talked about some time ago.

That would mean that you wouldn't have a chance to do non-mapper-based logic e.g.:
https://github.com/gradle/gradle/blob/2a32ceedf4406e782b697af9b396f2fcea7e7f9d/subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java#L151-L163
Currently, if you need some custom logic (something, which doesn't fit the mappers' concepts), you could just write it in this method; as it's used recursively, you wouldn't need to crunch those custom parts into mappers.

Reply:
That makes sense, thanks for the explanation.